sdk/inc/azure/iot/az_iot_provisioning_client.h

```diff
@@ -423,6 +423,47 @@ AZ_NODISCARD az_result az_iot_provisioning_client_query_status_get_publish_topic
     size_t mqtt_topic_size,
     size_t* out_mqtt_topic_length);
 
+/**
+ * @brief Azure IoT Provisioning Client options for az_iot_provisioning_client_get_request_payload.
+ *        Not currently used.  Reserved for future use.
+ *
+ */
+struct az_iot_provisioning_client_payload_options;
+
+/**
+ * @brief Builds the optional payload for a provisioning request.
+ * @remark Use this API to build an MQTT payload during registration.
+ *         This call is optional for most scenarios.  Some service
+ *         applications may require #custom_payload_property specified during
+ *         registration to take additional decisions during provisioning time.
+ *         For example, if you need to register an IoT Plug and Play device you must
+ *         specify its model_id with this API via the #custom_payload_property
+ *         "{\"modelId\":\"" your_model_id "\"}".
+ *
+ * @param[in] client The #az_iot_provisioning_client to use for this call.
+ * @param[in] custom_payload_property __[nullable]__ Custom JSON to be added to this payload.
+ * Can be `NULL`.
+ * @param[in] reserved __[nullable]__ Reserved field for future options to this function.
+ * @param[out] mqtt_payload A buffer with sufficient capacity to hold the MQTT payload.
+ * If successful, contains a null-terminated string with the topic filter that needs
+ * to be passed to the MQTT client.
+ * @param[in] mqtt_payload_size The size, in bytes of \p mqtt_payload.
+ * @param[out] out_mqtt_payload_length __[nullable]__ Contains the string length, in bytes, of \p
+ * mqtt_payload. Can be `NULL`.
+ * @pre \p client must not be `NULL`.
+ * @pre \p reserved must be `NULL`
```

Comment:
Will this break our API, if in future we allowed it being non-null.... Whole point of adding this feild was to not break the contract. 

Note: precondition are also contracts.

Reply:
I'd personally prefer to have this just as extra safety guard though its not a make or break of coures.

RE breaking contract if we used this, I'd view this more as a "relaxing the contract" than breaking.  We used to not allow a NULL and now we would if we used this field.

That said, I think this is a great thing to discuss with broader Azure SDK team since I tend to defer on issues like this.

Reply:
Yea I think John might be right here (fwiw). I think if we went from "can be `NULL` or valid value" to "must be `NULL`" that would break contract. If we in the future let people add it, other's which had it as `NULL` wouldn't be broken. Should confirm with the C sdk team regardless

Reply:
> I'd view this more as a "relaxing the contract" than breaking. 

👍